streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBVersionedStoreTest.java

```diff
@@ -0,0 +1,435 @@
+        putToStore("k", null, BASE_TIMESTAMP);
+        putToStore("k", "v2", BASE_TIMESTAMP - 2);
+        putToStore("k", "v1", BASE_TIMESTAMP - 1);
+        putToStore("k", "v4", BASE_TIMESTAMP - 4);
+
+        verifyGetNullFromStore("k");
+        verifyTimestampedGetNullFromStore("k", BASE_TIMESTAMP);
+        verifyTimestampedGetValueFromStore("k", BASE_TIMESTAMP - 1, "v1", BASE_TIMESTAMP - 1);
+        verifyTimestampedGetValueFromStore("k", BASE_TIMESTAMP - 2, "v2", BASE_TIMESTAMP - 2);
+        verifyTimestampedGetValueFromStore("k", BASE_TIMESTAMP - 3, "v4", BASE_TIMESTAMP - 4);
+    }
+
+    @Test
+    public void shouldPutOlderNullWithNonNullLatest() {
+        putToStore("k", "v", BASE_TIMESTAMP);
+        putToStore("k", null, BASE_TIMESTAMP - 2);
+        putToStore("k", null, BASE_TIMESTAMP - 1);
+        putToStore("k", null, BASE_TIMESTAMP - 4);
+        putToStore("k", "v5", BASE_TIMESTAMP - 5);
+        putToStore("k", "v3", BASE_TIMESTAMP - 3);
+        putToStore("k", null, BASE_TIMESTAMP - 6);
+
+        verifyGetValueFromStore("k", "v", BASE_TIMESTAMP);
+        verifyTimestampedGetValueFromStore("k", BASE_TIMESTAMP, "v", BASE_TIMESTAMP);
+        verifyTimestampedGetNullFromStore("k", BASE_TIMESTAMP - 1);
+        verifyTimestampedGetNullFromStore("k", BASE_TIMESTAMP - 2);
+        verifyTimestampedGetValueFromStore("k", BASE_TIMESTAMP - 3, "v3", BASE_TIMESTAMP - 3);
+        verifyTimestampedGetNullFromStore("k", BASE_TIMESTAMP - 4);
+        verifyTimestampedGetValueFromStore("k", BASE_TIMESTAMP - 5, "v5", BASE_TIMESTAMP - 5);
+        verifyTimestampedGetNullFromStore("k", BASE_TIMESTAMP - 6);
+    }
+
+    @Test
+    public void shouldPutOlderNullWithNullLatest() {
+        putToStore("k", null, BASE_TIMESTAMP);
+        putToStore("k", null, BASE_TIMESTAMP - 2);
+        putToStore("k", null, BASE_TIMESTAMP - 1);
+        putToStore("k", null, BASE_TIMESTAMP - 4);
+        putToStore("k", "v3", BASE_TIMESTAMP - 3);
+        putToStore("k", "v5", BASE_TIMESTAMP - 5);
+        putToStore("k", null, BASE_TIMESTAMP - 6);
+
+        verifyGetNullFromStore("k");
+        verifyTimestampedGetNullFromStore("k", BASE_TIMESTAMP);
+        verifyTimestampedGetNullFromStore("k", BASE_TIMESTAMP - 1);
+        verifyTimestampedGetNullFromStore("k", BASE_TIMESTAMP - 2);
+        verifyTimestampedGetValueFromStore("k", BASE_TIMESTAMP - 3, "v3", BASE_TIMESTAMP - 3);
+        verifyTimestampedGetNullFromStore("k", BASE_TIMESTAMP - 4);
+        verifyTimestampedGetValueFromStore("k", BASE_TIMESTAMP - 5, "v5", BASE_TIMESTAMP - 5);
+        verifyTimestampedGetNullFromStore("k", BASE_TIMESTAMP - 6);
+    }
+
+    @Test
+    public void shouldPutRepeatTimestampAsLatest() {
+        putToStore("k", "to_be_replaced", BASE_TIMESTAMP);
+        putToStore("k", "b", BASE_TIMESTAMP);
+
+        verifyGetValueFromStore("k", "b", BASE_TIMESTAMP);
+        verifyTimestampedGetValueFromStore("k", BASE_TIMESTAMP, "b", BASE_TIMESTAMP);
```

Comment:
Should we verify that `"to_be_replaced"` was not move and thus we cannot query it with `BASE_TIMESTAMP - 1`?